### PR TITLE
Resolve picture_password sync collisions via PicturePasswordCollisionOperation

### LINKED ISSUE #14697: Add PicturePasswordCollisionOperation that resolves conflicts by reassigning the local record
❌ **This issue is not open for contribution. Visit <a href="https://learningequality.org/contributing-to-our-open-code-base/" target="_blank">Contributing guidelines</a>** to learn about the contributing process and how to find suitable issues.

## Overview
 
When two devices independently assign the same `picture_password` sequence to different learners and then sync, the incoming learner's `FacilityUser` record fails Morango deserialization due to a `unique_together` constraint violation on `("dataset", "picture_password")`. 

The record never appears in `kolibriauth_facilityuser` and the Morango Store is left in an inconsistent state. A new `PicturePasswordCollisionOperation` is needed that detects these failures, resolves the collision by reassigning the local pre-existing `FacilityUser`'s picture password, and retries deserialization so the incoming record lands cleanly.
  
**Complexity:** Medium
**Target branch:** develop

### Context
 Picture login is enabled for a facility when `FacilityDataset.picture_password_settings` is not null.

**How the collision happens:**
Picture passwords are stored as a dot-separated string on `FacilityUser.picture_password` with a `unique_together` constraint on `("dataset", "picture_password")`. When two devices independently assign the same sequence to different learners and then sync, Morango deserialization attempts to write both records to `kolibriauth_facilityuser`. The second write violates the constraint. Without targeted handling, the `FacilityUser` record never appears in the app database, and the Morango `Store` record is left with `dirty_bit=True`, a `deserialization_exception` of `django.db.utils.IntegrityError`, and possibly spurious `DeletedModels` or `HardDeletedModels` entries as artifacts of the failure.
 
**The correct approach (modify the local record):**
The operation should find the **local `FacilityUser`** already in the database that holds the conflicting picture password and reassign that user to a new picture password via `local_user.save()`. The incoming Store record's serialized data stays untouched. After the local user is moved out of the way, `_deserialize_from_store` retries the incoming record, which now deserializes cleanly because the unique constraint no longer conflicts. When the client later pushes its reassigned local user back to the remote, it receives a normal update with no collision. Resolution happens once, on one device.

**Sync operation pattern:**
The operation must inherit from `ReceiverDeserializeOperation` in `kolibri/core/auth/sync_operations.py` and be registered at low priority in `kolibri/core/auth/kolibri_plugin.py`. It calls `super().handle(context)` to let normal deserialization run first, then inspects what failed and corrects those records. Read `kolibri/core/auth/sync_operations.py` and `docs/backend_architecture/facility_syncing/index.rst` in full before implementing. Read the exam and lesson operations in `kolibri/core/exams/kolibri_plugin.py` as a contrast: those return `False` because they do not handle deserialization themselves. This operation is different because it calls `super().handle()` and must return a `transfer_statuses` constant, never `False`.
 
### The Change

Create a new class `PicturePasswordCollisionOperation` in `kolibri/core/auth/sync_operations.py` that inherits from `ReceiverDeserializeOperation`. Register it in `kolibri/core/auth/kolibri_plugin.py` and with low priority so that it runs last in the deserialization stage's operation list, so normal deserialization has already run before this operation executes. 
 
The `handle` method should do the following in order:
 
1. **Apply context guards.** Assert that `context.sync_session`, `context.transfer_session`, `context.filter`, and `context.is_receiver` are all present/truthy, in exactly the same way `ReceiverDeserializeOperation` does. Do not invent a different guard.
2. **Call `super().handle(context)`** to run normal deserialization. Capture the result.
3. **If the result is not `transfer_statuses.COMPLETED`, return it immediately.** Do not interfere with error or pending states from normal deserialization.
4. **Query the Morango Store for incoming records that failed deserialization due to the collision.** Filter by:
   - `partition__startswith=dataset_id` (obtained via `get_dataset_id(context)` from `kolibri/core/auth/sync_event_hook_utils.py`)
   - `model_name=FacilityUser.morango_model_name`
   - `dirty_bit=True`
   - `deserialization_exception=_INTEGRITY_ERROR_EXCEPTION` (which is `"django.db.utils.IntegrityError"`)
   - `last_transfer_session_id=context.transfer_session.id`
5. **If the query returns no records, return `transfer_statuses.COMPLETED`.**
6. **Extract the conflicting picture passwords from the incoming records.** Loop through the broken stores, deserialize each store's JSON with `FacilityUser.deserialize(json.loads(store.serialized), sync_filter=context.filter)`, and collect the `picture_password` values into a set.
7. **Query `FacilityUser` for local records holding those passwords.** Use `FacilityUser.objects.filter(dataset_id=dataset_id, picture_password__in=conflicting_passwords)`. These are the pre-existing local users whose passwords conflict with the incoming records. Querying the app model directly is correct because these users are already deserialized into the database. Do not scan Store records for this.
8. **If no local users are found, return `transfer_statuses.COMPLETED`.** This handles the edge case where the collision was not caused by a local user.
9. **Build the available sequences pool.** Check `are_picture_passwords_exhausted(dataset_id)`. If not exhausted, compute `get_all_valid_sequences(PICTURE_PASSWORD_SET) - get_assigned_sequences(facility)` and shuffle. These utilities are in `kolibri/core/auth/utils/picture_passwords.py`.
10. **Reassign each local user's picture password.** For each `FacilityUser` in the query results:
    - Set `picture_password` to `available_sequences.pop()` if available, else `None`
    - Call `local_user.save()` to persist the change to the database immediately
11. **Clear deserialization errors on the incoming broken Store records.** Update them to set `deserialization_error=None` and `deserialization_exception=None`. This allows `_deserialize_from_store` to retry them.
12. **Clean up DeletedModels and HardDeletedModels** for the broken store IDs. These are artifacts of the failed deserialization.
13. **Call `_deserialize_from_store(context.sync_session.profile, filter=context.filter)`** to retry deserialization. The incoming records still have `dirty_bit=True`, so they will be processed. The local users' passwords have been changed in the database, so the unique constraint will not conflict.
14. **Return `transfer_statuses.COMPLETED`.**
**Important: do not modify the incoming Store records' `serialized` field.** The incoming records keep their original picture passwords. The `dirty_bit` on incoming records stays `True` so `_deserialize_from_store` processes them. The local users' Store records will be updated through the normal Morango serialization flow triggered by `local_user.save()`.
 
Create unit tests in `kolibri/core/auth/test/test_picture_password_sync_operation.py` and an integration test in `kolibri/core/auth/test/test_morango_integration.py`. See the Testing section for specifics.

### How to Get There

To exercise the operation manually during development:
 
1. Run two Kolibri instances configured for peer sync.
2. On both devices, create learners in the same facility with picture login enabled, such that two learners on different devices are assigned the same `picture_password` sequence.
3. Sync the two devices (Device A initiates).
4. After sync completes, on Device A query `FacilityUser` and confirm:
   - Both learners exist in `kolibriauth_facilityuser`
   - Device A's original local learner has a **new** picture password (it was reassigned)
   - The learner that arrived from Device B has its **original** picture password
5. On Device B, confirm:
   - Both learners exist
   - Device B's local learner still has its original picture password (no collision occurred on Device B because Device A resolved the conflict during the pull and pushed back the reassigned record)
6. Query the Morango `Store` for the affected records on both devices and confirm `dirty_bit=False`, `deleted=False`, and no `DeletedModels` or `HardDeletedModels` entries.
To run the relevant tests:
```
pytest kolibri/core/auth/test/test_sync_operations.py
pytest kolibri/core/auth/test/test_morango_integration.py::PicturePasswordCollisionIntegrationTestCase
```

### Out of Scope
- Changes to `_deserialize_from_store` in Morango (it does not need modification)
- UI changes for surfacing collision resolution to the user
- Handling collisions between two incoming records (both from the remote, neither local)
- Learners who deserialized successfully but arrived with `picture_password=None` because picture login was not active on the sending device at the time of creation
- Assignment during new learner creation
- Triggering bulk assignment when the facility setting is first enabled
- Any frontend changes

#### General
- [ ] A new `PicturePasswordCollisionOperation` class exists in `kolibri/core/auth/sync_operations.py`, inheriting from `ReceiverDeserializeOperation` with `priority = -1`
- [ ] The operation is registered at low priority in `kolibri/core/auth/kolibri_plugin.py`
- [ ] The operation applies the same context guards as `ReceiverDeserializeOperation` (receiver only, correct stage)
- [ ] `super().handle(context)` is called first; the operation returns immediately for any result other than `COMPLETED`
- [ ] The Store query filters by `partition`, `model_name`, `dirty_bit`, `deserialization_exception`, and `last_transfer_session_id`
- [ ] The operation does **not** modify the incoming Store records' `serialized` field
- [ ] The operation queries `FacilityUser` (the app model) to find local users holding conflicting passwords
- [ ] The operation reassigns the local `FacilityUser`'s `picture_password` via `local_user.save()`
- [ ] After reassignment, the operation clears `deserialization_error` and `deserialization_exception` on the incoming broken Store records
- [ ] The operation cleans up `DeletedModels` and `HardDeletedModels` for the incoming store IDs
- [ ] `_deserialize_from_store` is called after the local users are reassigned, successfully deserializing the incoming records with their original passwords
- [ ] The operation returns `transfer_statuses.COMPLETED`
- [ ] If no broken stores are found, the operation returns `transfer_statuses.COMPLETED` without modifying anything
- [ ] If no local users match the conflicting passwords, the operation returns `transfer_statuses.COMPLETED` without modifying anything
- [ ] `are_picture_passwords_exhausted()` is called once before the loop; if exhausted, the local user's `picture_password` is set to `None`
- [ ] Colliding learners are never left absent from `kolibriauth_facilityuser`

### Testing

- [ ] **Unit tests** in `kolibri/core/auth/test/test_picture_password_sync_operation.py`:
  - Test that when a collision is detected, the **local `FacilityUser`** is reassigned a new `picture_password` (not `None`, not the original value). The incoming Store's `serialized` field remains unchanged. The incoming Store has `deserialization_error=None` and `deserialization_exception=None`.
  - Test that `DeletedModels` and `HardDeletedModels` entries for the incoming store are removed after correction.
  - Test that when `are_picture_passwords_exhausted()` returns `True`, the local `FacilityUser`'s `picture_password` is set to `None`.
  - Test that when no available sequences remain (empty set from `get_all_valid_sequences`), the local `FacilityUser`'s `picture_password` is set to `None`.
  - Test that when no broken stores are found, the operation returns `COMPLETED` without modifying anything.
  - Test that when `super().handle()` returns something other than `COMPLETED`, the operation returns that result immediately without modifying any records.
  - Test that Store records with a different `deserialization_exception` (not `IntegrityError`) are not touched.
  - Test that Store records with `dirty_bit=False` are not touched.
  - Test that when no local `FacilityUser` matches the conflicting password (edge case), the operation returns `COMPLETED` without modifying anything.
- [ ] **Integration test** `PicturePasswordCollisionIntegrationTestCase` in `kolibri/core/auth/test/test_morango_integration.py`:
  - Create a facility, a local `FacilityUser` with `picture_password="1.2.3"`, serialize into Store, then create a remote Store record for a different user also with `picture_password="1.2.3"`.
  - Run `_deserialize_from_store` to trigger the collision, confirm the remote store has `dirty_bit=True` and `deserialization_exception=_INTEGRITY_ERROR_EXCEPTION`.
  - Run the operation and confirm:
    - The **local user** has a new `picture_password` (not `"1.2.3"`, not `None`).
    - The **remote user** exists in `FacilityUser` with its original `picture_password == "1.2.3"`.
    - The remote Store has `dirty_bit=False`, `deserialization_exception=None`, `deserialization_error=None`, and no `DeletedModels`/`HardDeletedModels` entries.
  - Use `TransactionTestCase` because Morango's `_begin_transaction` uses transaction isolation incompatible with `TestCase`'s database wrapping.

### References
- `kolibri/core/auth/sync_operations.py` -- where `PicturePasswordCollisionOperation` should be created; contains `ReceiverDeserializeOperation` (the parent class)
- `docs/backend_architecture/facility_syncing/index.rst` -- operations pattern, stage structure, and `transfer_statuses` contract
- `kolibri/core/auth/kolibri_plugin.py` -- where the operation must be registered at low priority
- `kolibri/core/auth/sync_event_hook_utils.py` -- `get_dataset_id(context)` for scoping the Store query to the correct dataset
- `kolibri/core/auth/utils/picture_passwords.py` 
- `kolibri/core/auth/models.py`

## AI usage
🤖 This issue was written with AI assistance (Claude), under supervision, review and final edits by @LianaHarris360. The bilateral resolution bug was identified through iterative debugging with real device logs in a two-device sync setup. The issue description, acceptance criteria, and test specifications were drafted by Claude based on the debugging session context, then reviewed for accuracy. 🤖

### PR DESCRIPTION
**Pull request by rtibblesbot**
**Plan:** Add `PicturePasswordCollisionOperation` to resolve `unique_together` constraint violations on `picture_password` during Morango deserialization when two devices independently assign the same picture password to different learners.

- [x] Write the failing unit test file
- [x] Implement `PicturePasswordCollisionOperation` in `sync_operations.py`
- [x] Register the operation in `kolibri_plugin.py`
- [x] Write the integration test
- [x] Lint and final check

status

## Summary

When two devices independently assign the same `picture_password` sequence to different learners and sync, the incoming `FacilityUser` fails Morango deserialization with `IntegrityError`, leaving the remote learner permanently absent from `kolibriauth_facilityuser`. This adds `PicturePasswordCollisionOperation` to detect, resolve, and retry those failures without modifying the incoming serialized data.

## References

Closes learningequality/kolibri#14697

## Reviewer guidance

- `sync_operations.py` line 421: `local_user.save()` (not `save(update_fields=[...])`) is required so Morango's `_morango_dirty_bit` is persisted to the DB; using `update_fields` would silently skip the dirty bit and break re-serialization on the next sync cycle.
- `sync_operations.py` line 423: `broken_store_ids` is a lazy `values_list` queryset used as a subquery. The deletes at lines 425–426 execute before the `broken_stores.update()` at line 428 clears `deserialization_exception`, so the subquery still matches the correct broken records when the deletes run.
- `sync_operations.py` line 433: calls the private `_deserialize_from_store` Morango API to retry deserialization in-place after the local user's password is reassigned.

To test:
```
pytest kolibri/core/auth/test/test_picture_password_sync_operation.py
pytest kolibri/core/auth/test/test_morango_integration.py::PicturePasswordCollisionIntegrationTestCase
```

## AI usage

Implemented with Claude Code following a pre-approved plan using TDD (failing tests written first). Reviewed Morango source (`SyncableModel.save`) to verify a post-implementation change to `save(update_fields=[...])` would break dirty-bit persistence, and reverted it before pushing.

---
_@rtibblesbot's comments are generated by an LLM, and should be evaluated accordingly_

<details><summary>How was this generated?</summary>

- Ran pre-flight CI checks (lint, format, tests) and verified all pass
- Rebased onto the target branch and resolved any conflicts
- Reorganized commit history into clean, logical commits
- Audited the diff to ensure only issue-relevant files are changed
- Built PR body from the repository's PR template with evidence blocks

</details>

### CODE REVIEW

File: kolibri/core/auth/sync_operations.py [hunk 1]
```diff
@@ -338,3 +349,86 @@ def handle_remote_user(self, context, user_id):
...
+            picture_password__in=conflicting_passwords,
+        )
+
+        if not local_users.exists():
+            return transfer_statuses.COMPLETED
+
+        if are_picture_passwords_exhausted(dataset_id):
+            available_sequences = []
+        else:
+            facility = local_users.first().facility
```
Comment (LianaHarris360) [inline kolibri/core/auth/sync_operations.py]: Use `Facility.objects.get(dataset_id=dataset_id)` to fetch the facility here.

**Review by LianaHarris360 — CHANGES_REQUESTED** [review]

File: kolibri/core/auth/sync_operations.py (hunk 1 above)
Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py]: Fixed — replaced `local_users.first().facility` with `Facility.objects.get(dataset_id=dataset_id)` and added the `Facility` import. Also removed the now-unnecessary `select_related("facility")` from the `local_users` queryset.

**Review by LianaHarris360 — APPROVED** [review]
I manually tested this PR, running two Kolibri instances, creating a new learner on both and giving them the same picture password, then synced one device to the other. The output was correct, the picture password of the learner on the client device was reassigned and both devices synced both learners successfully.

https://github.com/user-attachments/assets/480a87ca-2f25-4944-826a-20ac8f5b0eb8

CC @bjester in case he has any other comments or changes to suggest, otherwise, this PR can be merged.

File: kolibri/core/auth/sync_operations.py [hunk 2]
```diff
@@ -338,3 +350,86 @@ def handle_remote_user(self, context, user_id):
...
+
+        if are_picture_passwords_exhausted(dataset_id):
+            available_sequences = []
+        else:
+            facility = Facility.objects.get(dataset_id=dataset_id)
+            available_sequences = list(
+                get_all_valid_sequences(PICTURE_PASSWORD_SET)
+                - get_assigned_sequences(facility)
+            )
+            random.shuffle(available_sequences)
```
Comment (bjester) [inline kolibri/core/auth/sync_operations.py]: What is the purpose of the shuffle?

File: kolibri/core/auth/sync_operations.py [hunk 3]
```diff
@@ -338,3 +350,86 @@ def handle_remote_user(self, context, user_id):
...
+            deserialization_error=None,
+            deserialization_exception=None,
+        )
+
+        DeletedModels.objects.filter(id__in=broken_store_ids).delete()
+        HardDeletedModels.objects.filter(id__in=broken_store_ids).delete()
+
+        _deserialize_from_store(context.sync_session.profile, filter=context.filter)
+
+        return transfer_statuses.COMPLETED
```
Comment (bjester) [inline kolibri/core/auth/sync_operations.py:435]: Was there an issue with just returning `PENDING`, instead of manually calling `_deserialize_from_store`?

File: kolibri/core/auth/sync_operations.py [hunk 4]
```diff
@@ -338,3 +350,86 @@ def handle_remote_user(self, context, user_id):
...
+            )
+            random.shuffle(available_sequences)
+
+        for local_user in local_users:
+            local_user.picture_password = (
+                available_sequences.pop() if available_sequences else None
+            )
+            local_user.save()
+
+        broken_store_ids = list(broken_stores.values_list("id", flat=True))
```
Comment (bjester) [inline kolibri/core/auth/sync_operations.py]: This does not need to be wrapped in a list. Arguably, wrapping in the list makes in less efficient than letting the Django pass it to the DB as a subquery.

File: kolibri/core/auth/sync_operations.py [hunk 5]
```diff
@@ -338,3 +350,86 @@ def handle_remote_user(self, context, user_id):
...
+            deserialization_exception=_INTEGRITY_ERROR_EXCEPTION,
+            last_transfer_session_id=context.transfer_session.id,
+        )
+
+        if not broken_stores.exists():
+            return transfer_statuses.COMPLETED
+
+        conflicting_passwords = {
+            json.loads(store.serialized).get("picture_password")
+            for store in broken_stores
```
Comment (bjester) [inline kolibri/core/auth/sync_operations.py]: Would be more efficient to only select the value that is needed here. i.e. `values_list('serialized', flat=True)`

File: kolibri/core/auth/test/test_picture_password_sync_operation.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+from kolibri.core.auth.models import FacilityUser
+from kolibri.core.auth.sync_operations import _INTEGRITY_ERROR_EXCEPTION
+from kolibri.core.auth.sync_operations import PicturePasswordCollisionOperation
+
+COLLISION_PASSWORD = "1.2.3"
+OTHER_EXCEPTION = "django.core.exceptions.ValidationError"
+
+
+def _make_context(transfer_session_id=None, is_receiver=True):
+    ctx = mock.MagicMock()
```
Comment (bjester) [inline kolibri/core/auth/test/test_picture_password_sync_operation.py]: Using a mock spec would ensure that we're not inadvertently mocking attributes or methods that do no actually exist on the context class.

File: kolibri/core/auth/test/test_morango_integration.py [hunk 7]
```diff
@@ -1678,3 +1686,142 @@ def test_issue_fixed_in_morango_pull_146(self, servers):
             .filter(**base_log_params)
             .exists()
         )
+
+
+class PicturePasswordCollisionIntegrationTestCase(TransactionTestCase):
```
Comment (bjester) [inline kolibri/core/auth/test/test_morango_integration.py]: This integration test isn't correct. The test should match the structure, format, and approach of other tests in this file. It should use `multiple_kolibri_servers` to test the actual syncing process between Kolibri servers.

**Review by bjester — CHANGES_REQUESTED** [review]
The biggest issue is that the Morango integration tests are 'faking it'. The integration tests should use the same testing methodology that the other morango integration tests use. Additionally, there are some small performance tweaks that I think are worthwhile and some questions about the implementation.

File: kolibri/core/auth/sync_operations.py (hunk 2 above)
Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py]: The shuffle ensures that when multiple local users are reassigned in the same collision event (two learners both holding conflicting passwords simultaneously), each gets a non-deterministic sequence rather than the same predictable first element every time. Added a comment to make this explicit.

File: kolibri/core/auth/sync_operations.py (hunk 3 above)
Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py:435]: Returning `PENDING` would eventually work — on retry Morango calls the operation again, `super().handle()` runs `_deserialize_from_store` internally, and since the errors are cleared it succeeds. But it has two drawbacks: (1) it relies on Morango's retry/backoff machinery rather than being an explicit single-step retry, and (2) it triggers a full second pass through all dirty records rather than only the broken ones. Calling `_deserialize_from_store` directly is the minimal, deterministic action: clear the collision, retry once, done. Happy to switch to `PENDING` if you feel that's the right contract for this operation.

File: kolibri/core/auth/sync_operations.py (hunk 4 above)
Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py]: Fixed. Reordered to delete DeletedModels/HardDeletedModels before calling `broken_stores.update()`, so the queryset is evaluated while it still matches the records and Django can pass it as a SQL subquery.

File: kolibri/core/auth/sync_operations.py (hunk 5 above)
Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py]: Fixed — switched to `values_list('serialized', flat=True)`.

File: kolibri/core/auth/test/test_picture_password_sync_operation.py (hunk 6 above)
Comment (rtibblesbot) [reply kolibri/core/auth/test/test_picture_password_sync_operation.py]: Fixed — added `spec=LocalSessionContext` from `morango.sync.context`. That class has `is_receiver` as a property and the slot-based attributes (`sync_session`, `transfer_session`, `filter`) that the operation actually accesses, so any inadvertent access to a non-existent attribute will now raise `AttributeError` at test time.

File: kolibri/core/auth/test/test_morango_integration.py (hunk 7 above)
Comment (rtibblesbot) [reply kolibri/core/auth/test/test_morango_integration.py]: Rewritten as a `multiple_kolibri_servers(2)` test under `MultipleServerTestCase`, with the `INTEGRATION_TEST` skip guard matching the rest of the file. The test creates colliding learners on two independent servers, runs a real sync, and verifies that both learners land on the receiver with distinct passwords.

File: kolibri/core/auth/sync_operations.py (hunk 3 above)
Comment (LianaHarris360) [reply kolibri/core/auth/sync_operations.py:435]: In the other PR, when `PENDING` was returned, the sync never ended. The `NetworkDeserializeOperation` was continuously reported as pending and nothing advanced it to "completed" based on the logs from my manual testing. 

Granted, this was also when the picture password collision was never actually resolved in the code, since the incoming user was always reassigned. But the sync forever loading was no longer an issue after returning `COMPLETED` and manually calling `_deserialize_from_store` in the previous PR.

Comment (bjester) [reply kolibri/core/auth/sync_operations.py:435]: Okay, we can leave this as is for this PR, but I plan to investigate this later

File: kolibri/core/auth/sync_operations.py (hunk 2 above)
Comment (bjester) [reply kolibri/core/auth/sync_operations.py]: The non-deterministic approach could make bugs harder to find, because of its stochastic nature. Your response doesn't indicate that it's actually required, since the 'same predictable first element every time' should not matter because the set is reduced to only the available sequences. The shuffle doesn't eliminate that chance either, since the shuffle could very well produce the same first element with some chance.

Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py]: Removed. The available sequences list is already filtered to unassigned passwords, so ordering is irrelevant and the shuffle added only non-determinism.

File: kolibri/core/auth/sync_operations.py (hunk 4 above)
Comment (bjester) [reply kolibri/core/auth/sync_operations.py]: Okay the reordering is good, but you still have this wrapped in `list(...)` which will immediately evaluate it and not use it as a subquery.

Comment (rtibblesbot) [reply kolibri/core/auth/sync_operations.py]: Fixed — removed the `list()` wrapper so Django uses the queryset as a subquery in the `filter(id__in=...)` calls.

**Review by bjester — APPROVED** [review]
